numpy/core/src/multiarray/refcount.c.src

```diff
@@ -14,46 +14,113 @@
 #include "iterators.h"
 
 #include "npy_config.h"
-
 #include "npy_pycompat.h"
 
-static void
-_fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype);
 
+/* Fill the data item with None before XDECREF, used to clear the array */
+#define FILL_WITH_NONE(temp) do {        \
+    PyObject *none = Py_None;            \
+    NPY_COPY_PYOBJECT_PTR(data, &none);  \
+    Py_INCREF(none);                     \
+    Py_XDECREF(temp);                    \
+    } while (0)
 
-/*NUMPY_API
- * XINCREF all objects in a single array item. This is complicated for
- * structured datatypes where the position of objects needs to be extracted.
- * The function is execute recursively for each nested field or subarrays dtype
- * such as as `np.dtype([("field1", "O"), ("field2", "f,O", (3,2))])`
+
+/**begin repeat
+ * #name = INCREF, XDECREF, CLEAR, VISIT#
+ * #inner_function = _PyArray_Item_INCREF, _PyArray_Item_XDECREF,
+ *                   _PyArray_Item_CLEAR, _PyArray_Item_VISIT#
+ * #is_visit_func = 0, 0, 0, 1#
+ * #item_func = Py_XINCREF, Py_XDECREF,, Py_VISIT#
+ * #set_to_none = 0, 0, 1, 0#
+ * #silence_errors = 0, 0, 1, 1#
+ * #context = "", "", "<ndarray.tp_clear>", "<ndarray.tp_traverse>"#
  */
-NPY_NO_EXPORT void
-PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+
+/* Use macros to achieve the different behaviour and function signatures */
+#if !@set_to_none@
+   #define HANDLE_ITEM_AT(data) @item_func@(*(data))
+#else
+    #define HANDLE_ITEM_AT(data) Py_XSETREF(*(data), Py_None)
+#endif
+
+
+#define HANDLE_ITEM_AT_UNALIGNED(data)          \
+    do {                                        \
+        PyObject *_tmp;                         \
+        NPY_COPY_PYOBJECT_PTR(&_tmp, data);     \
+        HANDLE_ITEM_AT(&_tmp);                  \
+        if (@set_to_none@) {                    \
+            NPY_COPY_PYOBJECT_PTR(data, &_tmp); \
+        }                                       \
+    } while (0)
+
+
+#if @is_visit_func@
+    #define CALL_INNER_FUNCTION(data, descr) \
+        @inner_function@(data, descr, visit, arg)
+    // #define ERR_RETURN PyErr_WriteUnraisable(Py_None); return 0
+#else
+    #define CALL_INNER_FUNCTION(data, descr) @inner_function@(data, descr)
+    // #define ERR_RETURN return -1;
```

Comment:
```suggestion
```
(also above). This file is the main annoyance. While it got quite a bit shorter by using the `c.src` the macros, etc. also added quite a bit of complexity.

Plus, I think there may have been some actual bug fixes here for which there are no test, but I do not remember what the fixes were :).

Reply:
you could do a `git blame` to see where it changed

Reply:
Oh, that was just me remembering to fix it up, it was my own change.